File: 404. Sum of Left Leaves.cpp
/*Find the sum of all left leaves in a given binary tree.

Example:

    3
   / \
  9  20
    /  \
   15   7

There are two left leaves in the binary tree, with values 9 and 15 respectively. Return 24.*/

/*这道题的本质是遍历二叉树，所以我们可以用 Level Order Traverse + queue，注意对左子叶的判断和处理*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumOfLeftLeaves(TreeNode* root) {
        if (!root) return 0;
        queue<TreeNode*> q;
        q.push(root);
        int ans = 0;
        while (!q.empty()){
            TreeNode* n = q.front();
            q.pop();
            if (n->left && !n->left->left && !n->left->right){
                ans += n->left->val;
            }
            if (n->left) q.push(n->left);
            if (n->right) q.push(n->right);
        }
        return ans;
    }
};


/*我们也可以用stack, 跟上面的解法几乎一样，只是把queue换成了stack，q.front()换成了s.top(), 实际上遍历的顺序不同*/
class Solution{
public:
    int sumOfLeftLeaves(TreeNode* root){
        if (!root) return 0;
        stack<TreeNode*> s;
        s.push(root);
        int ans = 0;
        while (!s.empty()){
            TreeNode* n = s.top();
            s.pop();
            if (n->left && !n->left->left && !n->left->right){
                ans += n->left->val;
            }
            if (n->left) s.push(n->left);
            if (n->right) s.push(n->right);
        }
        return ans;
    }
};

File: python/404. Sum of Left Leaves.py
"""
404. Sum of Left Leaves
Easy

Given the root of a binary tree, return the sum of all left leaves.

A leaf is a node with no children. A left leaf is a leaf that is the left child of another node.

 

Example 1:
Input: root = [3,9,20,null,null,15,7]
Output: 24
Explanation: There are two left leaves in the binary tree, with values 9 and 15 respectively.


Example 2:
Input: root = [1]
Output: 0
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution(object):
    def sumOfLeftLeaves(self, root):
        if not root:
            return 0
        if root.left and not root.left.left and not root.left.right:
            return root.left.val + self.sumOfLeftLeaves(root.right)
        else:
            return self.sumOfLeftLeaves(root.left) + self.sumOfLeftLeaves(root.right)


class Solution:
    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        res = 0
        stack = [(root, False)]
        while stack:
            node, isleft = stack.pop()
            if not node.left and not node.right and isleft:
                res += node.val
            if node.left:
                stack.append((node.left, True))
            if node.right:
                stack.append((node.right, False))
        return res